### PR TITLE
Streamline CMake Libraries

### PR DESCRIPTION
**Pull request by thedavidchu**
We want to be able to include library directories more easily without having to specify the full relative path.

- [x] Change sequential and parallel targets to libraries
    - Reference: https://cmake.org/cmake/help/latest/command/add_library.html
- [x] Make `src/common` into an interface library
    - Reference: https://cmake.org/cmake/help/latest/command/add_library.html#interface-libraries
- [x] In `src` and `test`, link library
    - Reference: https://cmake.org/cmake/help/latest/command/target_link_libraries.html
- [x] Let CMake know dependers also need CMake include libraries (idk what @arctic-marmoset said lol)
    - Reference: https://cmake.org/cmake/help/latest/command/target_include_directories.html

N.B. target_link_libraries isn't useful. Reference: https://cmake.org/cmake/help/latest/command/target_link_directories.html

### CODE REVIEW

**Comment by arctic-marmoset** [comment]
We could also just move all the headers to a top-level include dir:
```
root
|--include
|  |--parallel.hpp
|  `--sequential.hpp
`--src

File: src/parallel/CMakeLists.txt [hunk 1]
```diff
@@ -1,10 +1,18 @@
 # NOTE: We include header files to make them visible to IDEs.
-add_executable(parallel
-    main.cpp
+add_library(parallel
     parallel.cpp
     parallel.hpp
 )
 
+add_executable(parallel_exe
```
Comment (arctic-marmoset) [inline src/parallel/CMakeLists.txt:7]:
```suggestion
# Forward this directory to dependents.
target_include_directories(parallel
    PUBLIC
    ${CMAKE_CURRENT_SOURCE_DIR}
)

add_executable(parallel_exe
```

**Comment by thedavidchu** [comment]
> We could also just move all the headers to a top-level include dir:
> 
> ```
> root
> |--include
> |  |--parallel.hpp
> |  `--sequential.hpp
> `--src
> ```

We decided not to do this but `target_include_directories` may be useful?

**Comment by arctic-marmoset** [comment]
The cleanest way to do this is like so:
```
src
|--common
|  |--include
|  |  `--common
|  |     |--logger.hpp
|  |     |--status.hpp
|  |     ...
|  `--CMakeLists.txt
|
|--parallel
|  |--include
|  |  `--parallel
|  |     `--parallel.hpp
|  |--parallel.cpp
|  `--CMakeLists.txt
...
```

This way, when you do
```cmake
target_include_directories(parallel
    PUBLIC
    include
)
```

That means you're adding `parallel/include/...` to the additional include paths. I.e., dependents will have access to `parallel/...` when they try to #include.

**Pull request merged**